doc/tutorials/expert/custom_components.md

```diff
@@ -0,0 +1,171 @@
+
+colors = {
+    "O": [0, 0, 0, 255],
+    "X": [247, 82, 0, 255],
+    " ": [247, 186, 119, 255],
+}
+
+# fmt: off
+box = [
+    ['O', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', 'O', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', 'O', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', 'X', 'X', 'X', 'X', 'X', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', 'X', 'X', 'O', 'O', 'O', 'X', 'X', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', 'X', 'X', 'O', ' ', ' ', 'X', 'X', 'O', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', 'X', 'X', 'O', ' ', ' ', 'X', 'X', 'O', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', 'O', 'O', ' ', 'X', 'X', 'X', 'O', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', 'X', 'X', 'O', 'O', 'O', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', 'X', 'X', 'O', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', ' ', 'O', 'O', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', 'X', 'X', ' ', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', 'X', 'X', 'O', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['X', ' ', 'O', ' ', ' ', ' ', ' ', ' ', 'O', 'O', ' ', ' ', ' ', 'O', ' ', 'O'],
+    ['X', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', 'O'],
+    ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O'],
+]
+# fmt: on
+
+np_box = np.array([[colors[c] for c in row] for row in box], dtype=np.uint8)
+np_box_as_list = [[[int(z) for z in y] for y in x] for x in np_box.tolist()]
+
+class MarioButton(JSComponent):
+
+    _esm = "mario_button.js"
+    _stylesheets = ["mario_button.css"]
+
+    _box = param.List(np_box_as_list)
+    gain = param.Number(0.1, bounds=(0.1, 1.0), step=0.1)
+    duration = param.Number(1.0, bounds=(0.5, 2), step=0.5,)
+    size = param.Integer(100, bounds=(10, 1000), step=10)
+    animate = param.Boolean(True)
+
+    margin = param.Integer(10)
+
+if pn.state.served:
+    button = MarioButton()
+    parameters = pn.Param(
+        button, parameters=["gain", "duration", "size", "animate"]
+    )
+    settings=pn.Column(parameters, "Credits: Trevor Mantz")
+    pn.FlexBox(settings, button).servable()
+```
+
+:::
+
+:::{dropdown} `mario_button.js`
+
+```javascript
+/**
+ * Makes a Mario chime sound using web `AudioContext` API.
+ *
+ * @see {@link https://twitter.com/mbostock/status/1765222176641437859}
+ *
+ */
+function chime({ gain, duration }) {
+    let c = new AudioContext();
+    let g = c.createGain();
+    let o = c.createOscillator();
+    let of = o.frequency;
+    g.connect(c.destination);
+    g.gain.value = gain;
+    g.gain.linearRampToValueAtTime(0, duration);
+    o.connect(g);
+    o.type = "square";
+    of.setValueAtTime(988, 0);
+    of.setValueAtTime(1319, 0.08);
+    o.start();
+    o.stop(duration);
+  }
+
+  /**
```

Comment:
We should get the types working.

Reply:
What does it mean to you to "get the types working" ?

Reply:
I just did not understand how to define types in javascript files. Thought they needed something special.

I will add a small section describing how to add them. That will give users intellisense in their editor making everything easier.